### ISSUE TITLE
Use indentation for delimiter mismatch detecting may get wrong help message

### ISSUE THREAD

**Issue by chenyukang** [rust-lang/rust#103882]
[For code](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=e1bf629a2cd5f612f3fedd2886a972ff) 

```rust
#![feature(let_chains)]
trait FnCtxt {}

impl dyn FnCtxt {
    pub fn report(&self) -> u32 {
        let sum = |a: u32,
                   b: u32,
                   c: u32| {
            a + b + c
        };
        sum(1, 2, 3)
    }

    fn check(&self, val: Option<u32>, num: Option<u32>) {
        if let Some(b) = val
        && let Some(c) = num {
        && b == c {
        }
    }
}

fn main() { }
```

Current output is:
```console
error: this file contains an unclosed delimiter
  --> src/main.rs:22:15
   |
4  | impl dyn FnCtxt {
   |                 - unclosed delimiter
...
8  |                    c: u32| {
   |                            - this delimiter might not be properly closed...
9  |             a + b + c
10 |         };
   |         - ...as it matches this but it has different indentation
...
22 | fn main() { }
   |               ^

```

It's error because indentation mismatch at line 8 and 10, ideally, we should report there is an extra `{` at line 16.

Or we only show line 4~22 has `unclosed delimiter`, wrong help message may mislead user.

The root cause is we use indentation width for checking possible position:
https://github.com/rust-lang/rust/blob/a876a4df32b402e3886cd9f2af02cff3dd8e21c8/compiler/rustc_parse/src/lexer/tokentrees.rs#L112

**Comment by chenyukang** [comment]
Related issue: https://github.com/rust-lang/rust/issues/68987

**Comment by chenyukang** [comment]
For code:
![image](https://user-images.githubusercontent.com/230646/200123582-d88f5853-7db1-4ba8-8666-412552763d71.png)

The line at 68 will also detected as not match, because of the `.`

**Comment by chenyukang** [comment]
Generally, I think in practice, use indentation width for detecting possible mismatch position is not good. 

Give a wrong tip will make user spent more time to find the root cause.
